lib/Serialization/Deserialization.cpp

```diff
@@ -6359,8 +6367,17 @@ void ModuleFile::finishNormalConformance(NormalProtocolConformance *conformance,
           "serialized conformances do not match requirement signature",
           llvm::inconvertibleErrorCode()));
     }
-    while (conformanceCount--)
-      reqConformances.push_back(readConformance(DeclTypeCursor));
+    while (conformanceCount--) {
+      auto conformance = readConformanceChecked(DeclTypeCursor);
+      if (conformance) {
+        reqConformances.push_back(conformance.get());
+      } else if (getContext().LangOpts.EnableDeserializationRecovery) {
+        consumeError(conformance.takeError());
+        reqConformances.push_back(ProtocolConformanceRef::forInvalid());
```

Comment:
An extra recovery here makes sense with the one above.

What is the error in normal compilation when this invalid conformance is passed down again?

Reply:
I'll add the ObjC test and have a look :)

Reply:
Okay, so turns out this needs some protocol and associated type fun to reproduce, not Objc. But I've got both now anyway! See the updated test.

I couldn't hit just this case without it crashing in the witness substitution (when not allowing errors). When allowing errors this ends up showing up as a conformance diagnostic:
```
error: referencing static method 'thing' on 'ReqProto' requires that 'EnumB.Value' conform to 'ProtoA'
```

Which is about as helpful as the missing type diagnostics:
```
error: cannot find type 'ProtoB' in scope
```